receiver/mongodbreceiver/client.go

```diff
@@ -0,0 +1,222 @@
+func (c *mongodbClient) Disconnect(ctx context.Context) error {
+	if c.isConnected(ctx) {
+		c.logger.Info(fmt.Sprintf("disconnecting from mongo connection for host(s): %v", c.hosts))
+		return c.driver.Disconnect(ctx)
+	}
+	return nil
+}
+
+// ListDatabaseNames gets a list of the database name given a filter
+func (c *mongodbClient) ListDatabaseNames(ctx context.Context, filters interface{}, options ...*options.ListDatabasesOptions) ([]string, error) {
+	return c.driver.ListDatabaseNames(ctx, filters, options...)
+}
+
+// Query executes a query against a database. Relies on connection to be established via `Connect()`
+func (c *mongodbClient) RunCommand(ctx context.Context, database string, command bson.M) (bson.M, error) {
+	db := c.driver.Database(database)
+	result := db.RunCommand(ctx, command)
+
+	var document bson.M
+	err := result.Decode(&document)
+	return document, err
+}
+
+// ServerStatus returns the result of db.runCommand({ serverStatus: 1 })
+// more information can be found here: https://docs.mongodb.com/manual/reference/command/serverStatus/
+func (c *mongodbClient) ServerStatus(ctx context.Context, database string) (bson.M, error) {
+	return c.RunCommand(ctx, database, bson.M{"serverStatus": 1})
+}
+
+// DBStats returns the result of db.runCommand({ dbStats: 1 })
+// more information can be found here: https://docs.mongodb.com/manual/reference/command/dbStats/
+func (c *mongodbClient) DBStats(ctx context.Context, database string) (bson.M, error) {
+	return c.RunCommand(ctx, database, bson.M{"dbStats": 1})
+}
+
+type mongoTestConnectionResponse struct {
+	Version string
+}
+
+// TestConnection validates the underlying connection of the client and returns a result of the
+// version of mongo the client is connected to so adjustments in collection protocol can
+// be determined
+func (c *mongodbClient) TestConnection(ctx context.Context) (*mongoTestConnectionResponse, error) {
+	if err := c.Connect(ctx); err != nil {
+		return nil, fmt.Errorf("unable to connect: %w", err)
+	}
+
+	res, err := c.RunCommand(ctx, "admin", bson.M{"buildInfo": 1})
+	if err != nil {
+		return nil, fmt.Errorf("unable to get build info: %w", err)
+	}
+
+	v, ok := res["version"].(string)
+	if !ok {
+		return nil, errors.New("unable to parse mongo version from server")
+	}
+
+	c.logger.Debug(fmt.Sprintf("retrieved version information from mongo server to be: %s", v))
+	return &mongoTestConnectionResponse{
+		Version: v,
+	}, nil
+}
+
+func (c *mongodbClient) ensureClient(ctx context.Context) error {
+	if !c.isConnected(ctx) {
+		driver, err := mongo.Connect(ctx, c.options())
+		if err != nil {
+			return fmt.Errorf("error creating mongo client: %w", err)
+		}
+
+		if err := driver.Ping(ctx, readpref.PrimaryPreferred()); err != nil {
+			return fmt.Errorf("could not connect to hosts %v: %w", c.hosts, err)
+		}
+
+		c.logger.Info(fmt.Sprintf("Mongo connection established to hosts: %s", strings.Join(c.hosts, ", ")))
+		c.driver = driver
+	}
+	return nil
```

Comment:
In Go, it's preferable to bail early in the simple case. This keeps the code flatter.

```suggestion
	if c.isConnected(ctx) {
		return nil	
	}
	
	driver, err := mongo.Connect(ctx, c.options())
	if err != nil {
		return fmt.Errorf("error creating mongo client: %w", err)
	}

	if err := driver.Ping(ctx, readpref.PrimaryPreferred()); err != nil {
		return fmt.Errorf("could not connect to hosts %v: %w", c.hosts, err)
	}

	c.logger.Info(fmt.Sprintf("Mongo connection established to hosts: %s", strings.Join(c.hosts, ", ")))
	c.driver = driver
	return nil
```